src/GraphQL/Internal/Validation.hs

```diff
@@ -386,12 +472,77 @@ resolveFragmentDefinitions allFragments =
           modify (Set.insert name)
           FragmentSpread name directives <$> resolveFragment' definition
 
+-- * Variables
+
+data VariableDefinition = VariableDefinition Variable AST.Type (Maybe Value) deriving (Eq, Ord, Show)
```

Comment:
No, the `Value` is a default value, so `Maybe Value` means you don't have to supply a default value. The AST module has a type alias from `Value` to `DefaultValue`—we can do that here if you think it would help.

I agree it's a bit of a shambles in Aeson. To avoid that, we need to be careful with defining `Maybe` instances for `FromValue` or `ToValue` 

Whether or not the value is nullable is encoded in `AST.Type`. You can see an example of how the query document defines the type here: https://facebook.github.io/graphql/#sec-Language.Variables and an example of defaults at https://facebook.github.io/graphql/#sec-Variable-Default-Values-Are-Correctly-Typed.

FWIW, the semantics of "no default value supplied" are that if the user doesn't provide a value and the value is nullable, use `nil`, if it's _not_ nullable, return an error.

This PR implements those semantics in `substituteVariables` but makes no other use of `AST.Type`.